### DISCUSSION TITLE
Welcome to floor Discussions!

### DISCUSSION THREAD

**Discussion by vitusortner** [vitusortner/floor#441] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by salvaterra** [comment]
Hi, 

My name is Roddy thanks for the library - very helpful!

What do you think about adding this to the main readme.md ? I think people will appreciate going from 30 lines to 2 for each DAO.

```
abstract class GenericDao<T> {

  @insert
  Future<void> insertObj(T obj);

  @insert
  Future<void> insertObjs(List<T> objs);

  @update
  Future<void> updateObj(T obj);

  @update
  Future<void> updateObjs(List<T> objs);

  @delete
  Future<void> deleteObj(T obj);

  @delete
  Future<void> deleteObjs(List<T> objs);

  @Query('SELECT * FROM \$DB_TABLE WHERE id = :id')
  Future<T> findObjById(int id);

  @Query('SELECT * FROM \$DB_TABLE')
  Future<List<T>> findAllObjs();

  @Query('SELECT * FROM \$DB_TABLE')
  Stream<List<T>> findAllObjsAsStream();
}

@dao
abstract class CatDao extends GenericDao<Cat>{
  String DB_TABLE = 'Cat';
}
```

**Reply by vitusortner** [reply]
Hi @salvaterra! We have a [section in the project's README](https://github.com/vitusortner/floor#inheritance-1) that briefly mentions the capabilities of using abstract classes as base DAOs. If you feel like we could improve that section, feel free to open a pull request.

**Comment by TalebRafiepour** [comment]
How to write migrations when I increase the database version?
Is there any example? How can I reach that?
Thanks for answering me
@vitusortner

**Reply by dkaera** [reply]
Hi @TalebRafiepour 
To run migration you need to increase the database version and add migration realization (from [oldVersion] to [newVersion] version). Here is an example https://github.com/vitusortner/floor/blob/7888578cda7e8d1982ff23d99e03334256768cd7/floor/test/adapter/migration_adapter_test.dart

**Reply by vitusortner** [reply]
You can find another example here https://vitusortner.github.io/floor/migrations/

**Reply by TalebRafiepour** [reply]
Thank you for your good support🌹🙏🙏🙏

**Comment by purplegamba** [comment]
Could someone please give me an example of how to use the 'unique' attribute in an entity?  It's mentioned only very briefly in the documentation.

**Reply by purplegamba** [reply]
Never mind, I think I figured it out.  Hopefully this is right:

@Entity(tableName: 'person', indices: [Index(value: ['custom_name'], unique: true)])
class Person {
  @primaryKey
  final int id;

  @ColumnInfo(name: 'custom_name', nullable: false)
  final String name;

  Person(this.id, this.name);
}

**Comment by ErliSoares** [comment]
First of all thank you very much for the library you are congratulating, I am very in doubt which library to use moor or floor, but I decided not to use any of them for now and use my custom one for now.

I just wanted to give feedback because I decided not to use this library for the time being.
* Not in production
* Use sqflite not [sqlite3](https://pub.dev/packages/sqlite3)
* Moor library is having more coding frequency
* We do not have support for WEB, I even managed to use sqflite on the WEB, [but there is no way to use an existing database for example](https://github.com/deakjahn/sqflite_web/issues/14)

**Reply by vitusortner** [reply]
Thanks for your feedback. We're continuously improving the library.

Floor's first stable release is just around the corner. Once our dependencies have migrated to null-safe Dart, we're good to go.

**Comment by xocechka** [comment]
I'm having trouble with Foreign Key action's
I started with the basic example Person - Dog. I cant delete a Person who has a Dog associated. Sorry for the English.

**Reply by vitusortner** [reply]
Can you paste a code snippet to find out where the problem lies?

**Comment by Tedesc0** [comment]
First of all, thanks for the framework. I'm using it in a couple of projects an it is very helpful.  I'd like to ask you to make the floor accept @required in the parameters for the DAO methods. I fought an error for a long time now and I think if the parameters were @required the problem haven't occurred.  Thanks in advance and sorry for the english.

**Reply by vitusortner** [reply]
Great to hear about your interest in the library! We're tracking this problem in a separate issue (#371), which you could subscribe to, to be notified once named/positional parameters are supported.

**Reply by Tedesc0** [reply]
@vitusortner, pls, have you received any reports that the floor has problems in test units with SDK 2.12? In my case, all functions  of a DAO that read from the database return null when called inside an expect() . To reproduce the problem I created a small piece of code that calls to DAO like as follows: 

```dart
  group('persistence agent', () {
    test('user persistence agent', () async {
      IUser? user;
      user = null;
      user = await persistenceManager.usersPersistence.readById(id: UserId.from(id: 'USER-------0'));
      print(user);
      user = null;
      expect(() async {
        user = await persistenceManager.usersPersistence.readById(id: UserId.from(id: 'USER-------0'));
      }, returnsNormally);
      print(user);
      user = null;
      user = await persistenceManager.usersPersistence.readById(id: UserId.from(id: 'USER-------0'));
      print(user);
    });
  });
```

The response at the console is:

USER-------0
null
USER-------0

The calls inside and outside the expect() test function are exactly the same. That's why I don't understand the difference shown in the console.

Could you help me in this subject?  Thanks a lot, in advance.

**Reply by mqus** [reply]
The function in `expect` is an async function, which means that it could be called later, e.g. after print(user) or even at the end of the function. I don't know if this happens here, you should look how expect is documented exactly. I can also imagine that there is shadowing going on, but I haven't tried it out.

**Reply by vitusortner** [reply]
Yeah, `expect` behaving differently than expected sounds likely to me here too. @Tedesc0 what do you need the async block in async for?

**Reply by Tedesc0** [reply]
@vitusortner ,I create unit tests for all database operations defined in the Persistence, Services and Repository layers of my applications and the operations can reach hundreds. Doing all tests without expect statement is very laborious and takes a lot of time. As database operations are asynchronous all methods in all classes of those layers are also asynchronous. Only to see if the method ended without throwing an error it is necessary to write a try-catch for each of them and this takes time. I expect to have answered your question.  Thanks a lot to reply to my post.

**Reply by mqus** [reply]
If you just want to check if operations run through without throwing an error I think you don't need `expect` or `try` at all, since the testing framework should already catch all errors and fail the test if an error occurs. As [stackoverflow says](https://stackoverflow.com/questions/65853973/in-dart-how-to-verify-a-void-method-that-does-not-throw-exception-in-unit-tests) the `returnsNormally` Matcher is only for nicer error messages.

PS: see also https://api.flutter.dev/flutter/test_api/throwsA.html for Futures. `returnsNormally` does not have the same documentation but I expect similar behaviour.

**Reply by mqus** [reply]
you probably want to use [expectLater](https://pub.dev/documentation/test_api/latest/test_api/expectLater.html):
```dart
[...]
await expectLater(() async {
        user = await persistenceManager.usersPersistence.readById(id: UserId.from(id: 'USER-------0'));
      }, returnsNormally);
[...]
```
I haven't tried it out but it seems to match your expectations

**Reply by Tedesc0** [reply]
> you probably want to use [expectLater](https://pub.dev/documentation/test_api/latest/test_api/expectLater.html):
> 
> ```dart
> [...]
> await expectLater(() async {
>         user = await persistenceManager.usersPersistence.readById(id: UserId.from(id: 'USER-------0'));
>       }, returnsNormally);
> [...]
> ```
> 
> I haven't tried it out but it seems to match your expectations

@mqus i already tried expectLater() and it also shows the same problem. The strangest thing in this situation is that expect () works OK for the INSERT of rows in the database, but does not work when SELECTing these same rows. As we used to say in Portuguese, "I am more lost than a blind man in a shooting". Thanks a lot for your interest in this problem.

**Comment by fillsinger** [comment]
Hi @vitusortner 

is it possible to model a many-to-many relationship in an entity as in Hibernate for example?
If not, how is it possible to handle many to many relationships?
e.g.
```dart
class Emplyoee {
  final int id;
  final String name;
  final List<Project> listofProjects;
}
```

**Reply by vitusortner** [reply]
Hi! Floor doesn't provide direct mapping for m-n relationships. You'll need to manually create an association table to map the relationships. A ticket for the requested functionality can be found here https://github.com/vitusortner/floor/issues/47.
